src/core/utilities/bucketlist.hpp

```diff
@@ -0,0 +1,451 @@
+            bucket_idx--;
+            curr = bucket_size - 1;
+        }
+        else
+        {
+            auto check
+                = bucket_idx >= 0
+                  and bucket_idx < static_cast<typename BucketListIndex::Bindex>(buckets_.size())
+                  and curr >= 0 and (curr - 1 < static_cast<BucketListIndex::Bindex>(bucket_size));
+            if (!check)
+            {
+                std::cout << curr << " " << bucket_idx << "\n";
+                throw std::runtime_error(
+                    "BucketListError : bucket_idx (" + std::to_string(bucket_idx) + "), curr ("
+                    + std::to_string(curr) + " bucket_size(" + std::to_string(bucket_size) + ")");
+            }
+            buckets_[bucket_idx][curr - 1] = BucketListIndex::default_idx_v;
+            curr--;
+        }
+    }
+
+    using bucket_t                              = std::array<std::size_t, bucket_size>;
+    typename BucketListIndex::Bindex bucket_idx = 0;
+    typename BucketListIndex::Bindex curr       = 0;
+    std::vector<bucket_t> buckets_;
+};
+
+
+// ----------------- BucketList Iterator ----------------------------------
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline auto& BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator++()
+{
+    curr_pos_++;
+    if (curr_pos_ == bucket_size)
+    {
+        curr_bucket_++;
+        curr_pos_ = 0;
+    }
+    return *this;
+}
+
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline auto& BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator--()
+{
+    if (curr_pos_ == 0)
+    {
+        curr_pos_ = bucket_size - 1;
+        --curr_bucket_;
+    }
+    else
+        --curr_pos_;
+    return *this;
+}
+
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline bool BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator!=(
+    bucket_iterator const& other) const
+{
+    return (other.curr_bucket_ != curr_bucket_ or other.curr_pos_ != curr_pos_)
+           or bucketsList_ != other.bucketsList_;
+}
+
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline auto& BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator+=(int n)
+{
+    auto globPos   = curr_pos_ + curr_bucket_ * bucket_size;
+    auto newGloPos = globPos + n;
+    curr_bucket_   = newGloPos / static_cast<int>(bucket_size);
+    curr_pos_      = newGloPos - curr_bucket_ * bucket_size;
+    return *this;
+}
+
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline auto BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator+(int n) const
+{
+    auto copy{*this};
+    copy += n;
+    return copy;
+}
+
+
+template<std::size_t bucket_size>
+template<typename BucketListPtr>
+inline auto BucketList<bucket_size>::bucket_iterator<BucketListPtr>::operator-(int n) const
+{
+    auto copy{*this};
+    auto globPos      = copy.curr_pos_ + copy.curr_bucket_ * bucket_size;
```

Comment:
`copy.curr_pos_ + copy.curr_bucket_ * bucket_size` 

could be a private funciton `globalPosition_()`, and replace a few other places it's calculated